llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -211,65 +382,88 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
-        FunctionCallee FC = M.getOrInsertFunction(SPIRVName, FT);
-        assert(FC.getCallee() && "SPIRV intrinsic creation failed");
-        CallInst *SPIRVCall =
-            CallInst::Create(FT, FC.getCallee(), Args, "", CI);
-        CI->replaceAllUsesWith(SPIRVCall);
-        // Mark the instruction with <symbolic_id, int_id> pair for later
+          Ins.first->second = NextID;
+        unsigned ID = Ins.first->second;
+
+        //  3. Transform to spirv intrinsic _Z*__spirv_SpecConstant* or
+        //  _Z*__spirv_SpecConstantComposite
+        SmallVector<unsigned, 4> GeneratedIDs;
+        auto *SPIRVCall = emitSpecConstantRecursive(ID, SCTy, GeneratedIDs, CI);
+        if (Ins.second) {
+          // emitSpecConstantRecursive might emit more than one spec constant
+          // (because of composite types) and therefore, we need to ajudst
+          // NextID according to the actual amount of emitted spec constants
+          NextID += GeneratedIDs.size();
+        }
+
+        if (IsComposite)
+          // __sycl_getCompositeSpecConstant returns through argument, so, the
+          // only thing we need to do here is to store into a memory pointed by
+          // that argument
+          new StoreInst(SPIRVCall, CI->getArgOperand(0), CI);
+        else
+          CI->replaceAllUsesWith(SPIRVCall);
+
+        // Mark the instruction with <symbolic_id, int_ids...> list for later
         // recollection by collectSpecConstantMetadata method.
-        setSpecConstMetadata(SPIRVCall, SymID, NextID - 1);
+        setSpecConstSymIDMetadata(SPIRVCall, SymID, GeneratedIDs);
         // Example of the emitted call when spec constant is integer:
         // %6 = call i32 @_Z20__spirv_SpecConstantii(i32 0, i32 0), \
         //                                          !SYCL_SPEC_CONST_SYM_ID !22
+        // !22 = {!"string-id", i32 0}
+        // Example of the emitted call when spec constant is vector consisting
+        // of two integers:
+        // %1 = call i32 @_Z20__spirv_SpecConstantii(i32 3, i32 0)
+        // %2 = call i32 @_Z20__spirv_SpecConstantii(i32 4, i32 0)
+        // %3 = call <2 x i32> @_Z29__spirv_SpecConstantCompositeii(i32 \
+        //          %1, i32 %2), !SYCL_SPEC_CONST_SYM_ID !23
+        // !23 = {!"string-id-2", i32 3, i32 4}
       } else {
         // 2a. Spec constant must be resolved at compile time - just replace
         // the intrinsic with default C++ value for the spec constant type.
-        CI->replaceAllUsesWith(getDefaultCPPValue(SCTy));
+        Value *Default = getDefaultCPPValue(SCTy);
+        if (IsComposite)
+          // __sycl_getCompositeSpecConstant returns through argument, so, the
+          // only thing we need to do here is to store into a memory pointed by
+          // that argument
+          new StoreInst(Default, CI->getArgOperand(0), CI);
+        else
+          CI->replaceAllUsesWith(Default);
```

Comment:
Nit
```suggestion
        if (IsComposite) {
          // __sycl_getCompositeSpecConstant returns through argument, so, the
          // only thing we need to do here is to store into a memory pointed by
          // that argument
          new StoreInst(Default, CI->getArgOperand(0), CI);
        } else {
          CI->replaceAllUsesWith(Default);
        }
```

Reply:
Fixed in 4c9ce329cecc0563d51fd71aa56696fa2ff40dba